design/027-package_types.md

```diff
@@ -0,0 +1,99 @@
+
+
+## Detailed Design
+
+Package recipes will contain a new attribute that defines the package type:
+
+```python
+class Pkg(ConanFile):
+    settings = "os"
+    package_type = "shared-library"
+```
+
+It is also possible to define such attribute in the ``configure()`` method:
+
+```python
+class Pkg(ConanFile):
+    def configure(self):
+        if something:
+            self.package_type = "shared-library"
+        else:
+            self.package_type = "application"
+```
+
+The initial proposal is to define the following package types:
+
+- "static-library": a regular static library (.lib, .a) with its public headers. To be defined only when a packaged library is always static (no "shared" option exists).
+- "shared-library": a regular shared library (.so, .dll, .dylib) with its public headers, and in the case of Windows, the typical small .lib linkage library. To be defined only when a packaged library is always shared (no "shared" option exists).
+- "header-library": a library containing exclusively headers, but not binary compiled artifacts. To be defined only when a packaged library is always shared (no "shared" option exists).
+- "library" : This is an abstract type, this type will be resolved to any of the 3 above cases based on options. The ``shared`` option must exist and have a value, otherwise, the recipe will fail. The optional ``header_only`` option would be used to define if it is a "header-library". This type is not necessary to be defined, but if it is defined, it will enforce the existence of a ``shared``.
+- "application": Executable artifacts, .exe, binaries to run. It will not contain headers, static libraries, or shared libraries to be linked (might contain shared libraries used by the executable, but exclusively for the executable). Not necessary to define it, but it might be informative for recipe readers to realize that such "conanfile.py" is creating an application and not a library.
+- "python-require": This is a pure python recipe, it actually doesn’t contain any package binary or artifact at all.
+- "assets": This package contains some assets, like images, sound files, videos, fonts, configuration files, etc. They are not C/C++ headers, not libraries to link or applications to run.
+- "unknown": If the type is not defined explicitly, or cannot be deduced from options
+
+This is just the initial proposal, please report on any other potential package types that you are already managing (better not to speculate, but build on top of actual use cases) and would benefit from an explicit model.
```

Comment:
We currently have a package that contains only CMake helper scripts. In Conan 1.X it's advertised as `header_only`, however, in this model, it would be either `assets` or `unknown`. I'm not sure that `assets` is the best name for such package class - maybe something like `scripts` or `build-scripts`?

I'm open to suggestions...

Reply:
Good case.

Is there something in the package that would condition how it is required? Would a ``build_requires`` be able to consume this package without issues?:
- header=False: No headers included
- libs=False: No libs to link
- build=True: Necessary at build time, in build context
- run=True: No need to put anything in the path, not an executable, not a shared lib. But it certainly need to be there at build time, or build will fail
- visible=False: Downstream consumers of the package are not aware how this package was built or even its build system

So maybe it is not necessary to explicitly classify it, as long as it can be consumed successfully without issues with the new ``requires`` model. If it needs to be typed, I'd probably prefer ``build-scripts``, to clearly indicate the intent.

Reply:
@memsharded as we agreed, add the `build-scripts` to the proposal. Introduced here: https://github.com/conan-io/conan/pull/10076